### PR TITLE
GitHub sponsor subscriptions

### PR DESCRIPTION
**Pull request by ErikSchierboom**
Add GitHub sponsor subscriptions

### CODE REVIEW

File: app/commands/donations/stripe/subscription/update_amount.rb [hunk 1]
```diff
@@ -25,9 +25,5 @@ def call
     )
 
     subscription.update!(amount_in_cents:)
-
-    # Update based on whether there is another different active subscription
-    user = subscription.user
-    User::SetActiveDonationSubscription.(user, user.donation_subscriptions.active.exists?)
```
Comment (ErikSchierboom) [inline app/commands/donations/stripe/subscription/update_amount.rb:31]: As we're only updating the amount of the subscription, `user.donation_subscriptions.active` doesn't change

File: app/controllers/webhooks/github_sponsors_controller.rb [hunk 2]
```diff
@@ -6,7 +6,11 @@ def create
       ProcessGithubSponsorUpdateJob.perform_later(
         # params[:action] does not work as it is populated by Rails with the action method name
         request.request_parameters[:action],
-        params[:sponsorship][:sponsor][:login]
+        params[:sponsorship][:sponsor][:login],
+        params[:sponsorship][:node_id],
+        params[:sponsorship][:privacy_level],
```
Comment (ErikSchierboom) [inline app/controllers/webhooks/github_sponsors_controller.rb]: I was wondering how we want to handle the privacy level. I can of course update the user's `show_on_supporters_page` property, but I might be overriding something that was already set (the field can't be  `null` so I can't know if the user changed this value, can I?)

File: db/migrate/20230418133624_change_external_receipt_url_to_null_in_donations_payments.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class ChangeExternalReceiptUrlToNullInDonationsPayments < ActiveRecord::Migration[7.0]
+  def change
+    change_column_null :donations_payments, :external_receipt_url, true
```
Comment (ErikSchierboom) [inline db/migrate/20230418133624_change_external_receipt_url_to_null_in_donations_payments.rb:3]: AFAICT there is no receipt URL for GitHub sponsorships

File: app/commands/user/set_active_donation_subscription.rb
Comment (ErikSchierboom) [inline app/commands/user/set_active_donation_subscription.rb:1]: I've refactored this to capture the logic of how to determine the active donation subscription (via `user.donation_subscriptions.active.exists?`) within the command, which means that the caller doesn't have to do that (there was some repeated logic). Then I renamed the command to `UpdateActiveDonationSubscription`.

File: app/jobs/process_github_sponsor_update_job.rb [hunk 4]
```diff
@@ -1,12 +1,19 @@
 class ProcessGithubSponsorUpdateJob < ApplicationJob
   queue_as :default
 
-  def perform(action, gh_username)
-    return unless action == 'created'
-
+  def perform(action, gh_username, node_id, privacy_level, is_one_time, monthly_price_in_cents)
     user = User.find_by(github_username: gh_username)
```
Comment (ErikSchierboom) [inline app/jobs/process_github_sponsor_update_job.rb:5]: Do we want to attempt to automatically update a user's github_username if we can't find it? We try to etch the user's email via another API call (it is not in the payload I think)

File: app/jobs/process_github_sponsor_update_job.rb [hunk 5]
```diff
@@ -1,12 +1,19 @@
...
+    case action
+    when 'cancelled'
+      Donations::Github::Sponsorship::HandleCancelled.(user, node_id, privacy_level, is_one_time, monthly_price_in_cents)
+    when 'created'
+      Donations::Github::Sponsorship::HandleCreated.(user, node_id, privacy_level, is_one_time, monthly_price_in_cents)
+    when 'edited'
+      Donations::Github::Sponsorship::HandleEdited.(user, node_id, privacy_level, is_one_time, monthly_price_in_cents)
+    when 'tier_changed'
+      Donations::Github::Sponsorship::HandleTierChanged.(user, node_id, privacy_level, is_one_time, monthly_price_in_cents)
+    end
```
Comment (ErikSchierboom) [inline app/jobs/process_github_sponsor_update_job.rb]: I'll cleanup unused parameters once we've decided what we want to do with the privacy level (see other comment)

File: app/commands/donations/github/sponsorship/handle_cancelled.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# This responds to a Github sponsorship webhook event with
+# the action being 'cancelled'
+class Donations::Github::Sponsorship::HandleCancelled
+  include Mandate
+
+  initialize_with :user, :node_id, :privacy_level, :is_one_time, :monthly_price_in_cents
```
Comment (iHiD) [inline app/commands/donations/github/sponsorship/handle_cancelled.rb]: We don't need/use `monthly_price_in_cents` or `privacy_level ` here.

File: app/commands/donations/github/sponsorship/handle_created.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# This responds to a Github sponsorship webhook event with
+# the action being 'created'
+class Donations::Github::Sponsorship::HandleCreated
+  include Mandate
+
+  queue_as :default
+
+  initialize_with :user, :node_id, :privacy_level, :is_one_time, :monthly_price_in_cents
```
Comment (iHiD) [inline app/commands/donations/github/sponsorship/handle_created.rb]: We don't need `privacy_level` here.

File: app/commands/donations/github/sponsorship/handle_edited.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# This responds to a Github sponsorship webhook event with
+# the action being 'edited'
+class Donations::Github::Sponsorship::HandleEdited
+  include Mandate
+
+  initialize_with :user, :node_id, :privacy_level, :is_one_time, :monthly_price_in_cents
```
Comment (iHiD) [inline app/commands/donations/github/sponsorship/handle_edited.rb]: Do we need to update the subscriptions' `monthly_price_in_cents` here?

Comment (iHiD) [reply app/commands/donations/github/sponsorship/handle_edited.rb]: We can just get rid of this?

File: app/commands/donations/github/sponsorship/handle_tier_changed.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# This responds to a Github sponsorship webhook event with
+# the action being 'tier_changed'
+class Donations::Github::Sponsorship::HandleTierChanged
+  include Mandate
+
+  initialize_with :user, :node_id, :privacy_level, :is_one_time, :monthly_price_in_cents
```
Comment (iHiD) [inline app/commands/donations/github/sponsorship/handle_tier_changed.rb]: Remove `privacy_level`

File: app/commands/donations/github/sponsorship/handle_cancelled.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
+# This responds to a Github sponsorship webhook event with
+# the action being 'cancelled'
+class Donations::Github::Sponsorship::HandleCancelled
+  include Mandate
+
+  initialize_with :user, :node_id, :is_one_time, :monthly_price_in_cents
```
Comment (iHiD) [inline app/commands/donations/github/sponsorship/handle_cancelled.rb]: Get rid of `monthly_price_in_cents`

**Review by iHiD — APPROVED** [review]

**Pull request merged**